articles/cross_compilation_build_tools_design.md

```diff
@@ -0,0 +1,251 @@
+Original Author: {{ page.author }}
+
+## Background
+
+The [cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation),
+and the [steve/ros2_raspbian_tools repository](https://github.com/esteve/ros2_raspbian_tools)
+contain instructions for cross-compiling ROS 2 to unsupported platforms like
+ARM64 and ARM-HF, using some configuration files for CMake and Docker. However,
+those instructions don't include how to build other ROS 2 packages, but just
+for building a whole ROS 2 distribution. Ideally, we should be able to
+cross-compile a ROS package for ROS 2 by launching a single `colcon build`
+command. Finally, even though other projects like [steve/ros2_raspbian_tools](https://github.com/esteve/ros2_raspbian_tools),
+[esteve/ros2_java](https://github.com/esteve/ros2_java) provide support
+for additional platforms like Android and iOS, we would like to have an extensible
+design that allows supporting new platforms with low effort.
+
+## Proposed Approach
+
+We propose continuing with the general approach of the cross-compilation
+tutorial, based on building a sysroot for the target platform using QEMU and
+Docker, and then using `cmake-toolchains` with a toolchain file pointing
+`CMAKE_SYSROOT` to the sysroot location to compile with C and C++
+cross-compilers.
+
+On top of that we propose using a **colcon plugin `cc-build`** to wrap the
+instructions on the cross-compilation tutorial into a single build command.
+That plugin would define a new `colcon_core.verb` entry point pointing to a
+class `CCBuildVerb` that uses the [Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern)
+to wrap the [`BuildVerb` class](https://github.com/colcon/colcon-core/blob/master/colcon_core/verb/build.py#L74)
+used by `colcon build` with additional functionality to handle the sysroot and
+toolchain file, and to inject additional `cmake-args` arguments. That would
+allows us to run the single command below from a ROS 2 overlay workspace on a
+developer workstation, for cross-compiling the packages in the workspace up to
+a package `performance_test` for the ARM-HF platform.
+
+```bash
+colcon cc-build --arch armhf --os ubuntu_bionic \
+  --packages-up-to performance_test # other `colcon build` arguments
+```
+
+That would should result on a new workspace directory `armhf-ubuntu_bionic`
+with subdirectories `build`, `install`, and `log` directories, with the
+ross-compiled binaries. All options for `colcon build` will be supported by
+forwarding them to the `BuildVerb` class. That would be the only command
+required for cross-compiling the packages in a workspace. Under the hood, the
+plugin would do the following:
+
+- The [development tools described in the cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation/#install-development-tools)
+  (`qemu-user-static`, `g++-aarch64-linux-gnu`, ...) should be installed in the
+  development workstation used for the build. The `cc-build` plugin will
```

Comment:
While doing the build within Docker sounds like a good idea I think the (to be developed?) tools should be not limited to that use case. Imo there should be "tools" which can do they job (in whatever environment) and then there can be a specific "meta tool" which uses the tools in a certain way - e.g. within a Docker container. Developing something only for a very use case will likely make the result not really reusable and be only useful to a small subset of users.

Reply:
Another proposal would be splitting this into two commands, as suggested in this thread several times:

- A new command `setup-sysroot` would use Docker to generate the sysroot
- A new colcon mixin that adds options for using the sysroot to the colcon build task. This command requires having the cross compilation toolchain installed.
- On an second iteration we might:
  - Maintain a docker image for cross compilation that has installed both the cross compilation toolchain, and the aforementioned colcon mixin, based on the images at the ros2/cross_compile repository.
  - Implement a new `cross-build` command (definitive name TBD) that will ensure the sysroot is created using `setup-sysroot`, and then launch the cross compilation into the docker container with the cross compilation toolchain.

How does that sound?

Reply:
> A new command `setup-sysroot` would use Docker to generate the sysroot

What exactly is this *verb* supposed to be doing? (Depending on the answer) why should this be a `colcon` verb?

> A new colcon mixin that adds options for using the sysroot to the colcon build task. This command requires having the cross compilation toolchain installed.

Sounds good to me.

Reply:
> What exactly is this verb supposed to be doing? (Depending on the answer) why should this be a colcon verb?

The idea is keeping a directory in the workspace for that target platform, for example  `armhf-ubuntu_bionic`, determined by convention based on arguments for architecture and OS. Let us call that directory the cc-root. The command `setup-sysroot` would create a `sysroot` subdirectory of the cc-root (e.g.  `armhf-ubuntu_bionic/sysroot`) with the sysroot for the target platform. The colcon mixin for cross-compilation would then use the same convention as `setup-sysroot` to determine the cc-root, and use the sysroot stored there. We'd have mixins that provide shortcuts for cross compiling known architecture-os combinations. 

Here I'm not proposing having `setup-sysroot` as a new colcon verb, but as a new separate command. I'm also open to use a different name instead of "setup-sysroot".

Reply:
> Here I'm not proposing having setup-sysroot as a new colcon verb, but as a new separate command.

Oh, outside of `colcon`? Please feel free to choose whatever you like - I don't have an opinion on that.

Reply:
I have updated the design to replace the colcon plugin by the new command and the colcon mixin, please let me know what you think

Reply:
Sounds reasonable to me. With this update I might not be the best person to provide more feedback since the `colcon` part of this boils down to a mixin. Maybe it would be good to update the title of the ticket to reflect the current state?

I am not sure I understand how the user gets the CMake toolchain file which would be referenced from the mixin. But that might be a too detailed question for now.